Add an sqlite based rpmdb backend (experimental)

pmatilai opened the pull request:
All normal functionality is expected to work. Automatic generation
of missing index tables is missing, but that's not relevant at this time.
Going forward, we'll also want some sort of compatibility tracking
for the sql schema.

The database scheme basically just mirrors what BDB does, using strings
for strings and blobs for everything else due to the way integers are
handled in the sqlite C API, for now at least.

Performance is similar or better with BDB in the current unsafe CDB
model, but sqlite uses proper database transactions so this is expected
to be an order of magnitude more robust.

Many things are stupid and/or kind of backwards here due to the internal
API, which I've avoided changing in order to keep it backportable for the
time being. https://github.com/rpm-software-management/rpm/pull/836 is
needed but otherwise this should drop quite trivially into 4.14.x too.
However as we're planning for a longer term future here, it would be dumb
to limit ourselves by what's possible with an internal BDB-oriented API,
so I've fairly major changes planned in that direction.

Note that this changes the default backend for the sake of testing
it in the CI, that will need to be removed before merging, the time to
talk about default changes is (much) later.

pmatilai commented:
For further background on the topic:

The world is full of key-value databases, but there are *very* few databases that support the kind of multiprocess concurrency that rpm requires, and even fewer that are license compatible. There's LMDB, but existing versions have severe limitations to key size (far below common PATH_MAX) which makes it unsuitable for rpm production use. Their upstream promised to up the limit over three years ago but it still hasn't happened and we ran out of time five years ago. There's something known as WiredTiger which is supposed to be the successor to BDB and according to spec sheet is technically suitable, but it's hardly time-proven, nobody has ever heard of it and its license is incompatible (GPLv3). From there on it gets even more exotic and incompatible.

And then there's sqlite, which obviously is not a key-value store but can serve the needs of rpm just fine. It's been there all these years, it's ubiquitous and has just the kind of track record we're looking for, but it's been overlooked because, well, we already had an sqlite backend and ripped it off, so there must be something wrong with it, right? There are couple of important factors here:
- The old sqlite backend in the 4.4.x days emulated bdb in a very literal way, using multiple separate whole databases for the index tables and all, so it couldn't do what it does best *at all*.
- The old sqlite backend was also limited concurrency-wise, because back then sqlite didn't have WAL support yet. These days, the amount of concurrency sqlite supports is *just* enough for rpm's needs.
- Third, the old sqlite backend did chroot in/out from the database code. That is what commit 0508e9a6e3311ed00c34887fd7715d3b469cdca6 is all about - to eliminate the need for that. In reality, BDB needs that just as much, it currently only works through the smallest of slightly lucky margins, and to do transactions on BDB level would require the same thing (which is something we could do now too, but that's another topic)

So in the end, going (back) to sqlite was not much of choice, it simply was the last database standing: no arbitrary key limitations, multiprocess concurrency, proven track record and stable file format, good performance, sane API etc. It also has one *huge* bonus over the key-value alternatives: you can actually inspect the database contents interactively while sitting in your lazy chair. So it's actually rather pleasant to work with. (however there's also the danger that it's *too* nice to work with and people will start poking directly into the sql instead of going through librpm)

mlschroe commented:
I wouldn't dismiss poor ndb that easily. It has not many LoC and fits rpm's needs perfectly. The packages database is similar to what rpm-3 had in the past (but written with fault tolerance in mind), and the indexes are just mmaped hash tables.

Conan-Kudo commented on lib/backend/dbi.c:
```diff
@@ -59,6 +64,15 @@ dbDetectBackend(rpmdb rdb)
 #endif
 
+#if defined(WITH_SQLITE)
+    path = rstrscat(NULL, dbhome, "/Packages.sqlite", NULL);
```
For this, we only have the single database file, right? Couldn't we just call this `rpmdb.sqlite` instead of `Packages.sqlite`?

pmatilai replied:
The thought has crossed my mind as well. The "reason" to have it resemble the others is basically to allow covering the different variants with "Packages*" but then LMDB doesn't follow that and we'll need an API to cover for the differences anyway so yeah, we could just as well call it something more sensible.

In general, naming (including database columns) is pretty much in the open still.

Conan-Kudo commented:
@mlschroe @pmatilai Well, @midipix and @Redfoxmoon3 are experimenting with the ndb backend for rpm-based [midipix](https://midipix.org/). It's been interesting...

I'm more interested in the SQLite backend simply because the tooling is better for diagnosing and repairing SQLite databases.

(cc: @dmach)

pmatilai commented:
@mlschroe , choice is good, right? ;)

There's value in having a minimal built-in solution, and there's value in having a solution based on proven standard building blocks, they just appeal to different audiences. And right now we don't really have the latter: BDB is a six years unmaintained behemoth nobody wants to touch with a ten foot pole, LMDB is three years waiting for upstream to up the key size, and we're stuck in the middle. This is intended to fill that particular gap.

pmatilai commented:
Changed database filename to rpmdb.sqlite, dropped the %_db_backend change to eliminate the merge blocker (which also means that this isn't getting tested, but can't be helped). 

Further development is to be expected plenty of it but I want to get this in before proceeding with other internal dbapi changes. It'll be more work per change that way of course, but then we'll have a backportable reference point, and having such a different player aboard should make the dbapi related improvements far more obvious.

mlschroe commented:
@Conan-Kudo What does "It's been interesting..." mean? Did they run into some problems?

Conan-Kudo commented:
@mlschroe In the beginning, it wasn't even workable because it needed memory map functions that weren't implemented yet. I had similar problems trying to use ndb for rpm on macOS. Midipix later implemented them, and we switched from LMDB to ndb. I'm thinking of using this new SQLite backend for rpm on macOS, since ndb isn't an option there.

It's been fairly difficult to debug issues with ndb because we can't easily tell what's going wrong or examine the contents of the rpmdb.

pmatilai commented:
@Conan-Kudo , @mlschroe - I'm interested in hearing about ndb usage experiences too, but please open a separate ticket (or use rpm-maint) for further discussion on that topic.
